### PR TITLE
sql: fast path for UPSERT

### PR DESCRIPTION
**Pull request by danhhz**
If a table has no secondary indexes and all columns are supplied, an UPSERT can
run without fetching previous values.

```
name                    old time/op    new time/op    delta
Upsert1_Cockroach-8       1.17ms ± 1%    1.00ms ± 0%  -14.64%  (p=0.008 n=5+5)
Upsert10_Cockroach-8      1.75ms ± 0%    1.46ms ± 1%  -16.63%  (p=0.016 n=4+5)
Upsert100_Cockroach-8     7.19ms ± 2%    5.55ms ± 2%  -22.82%  (p=0.008 n=5+5)
Upsert1000_Cockroach-8    71.9ms ± 5%    45.5ms ± 1%  -36.69%  (p=0.008 n=5+5)

name                    old alloc/op   new alloc/op   delta
Upsert1_Cockroach-8       86.2kB ± 0%    73.3kB ± 0%  -15.00%  (p=0.008 n=5+5)
Upsert10_Cockroach-8       169kB ± 0%     146kB ± 0%  -13.35%  (p=0.008 n=5+5)
Upsert100_Cockroach-8     1.07MB ± 0%    0.90MB ± 0%  -15.95%  (p=0.008 n=5+5)
Upsert1000_Cockroach-8    9.34MB ± 0%    7.69MB ± 0%  -17.68%  (p=0.008 n=5+5)

name                    old allocs/op  new allocs/op  delta
Upsert1_Cockroach-8        1.14k ± 0%     0.93k ± 0%  -18.05%  (p=0.029 n=4+4)
Upsert10_Cockroach-8       1.58k ± 0%     1.27k ± 0%  -19.71%  (p=0.016 n=4+5)
Upsert100_Cockroach-8      6.23k ± 0%     4.75k ± 0%  -23.74%  (p=0.008 n=5+5)
Upsert1000_Cockroach-8     52.0k ± 0%     39.3k ± 0%  -24.56%  (p=0.008 n=5+5)
```

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/6673)

### CODE REVIEW

**Comment by tamird** [comment]
How come this fast path can't be used if a table has secondary indexes? Isn't it true that only unique indexes are problematic?

**Comment by paperstreet** [comment]
In the update (conflict) case, the secondary index entries would have to be cleaned up, which you can't do without fetching the old data

**Comment by tamird** [comment]
Ah, makes sense.

<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> In the update (conflict) case, the secondary index entries would have to be cleaned up, which you can't do without fetching the old data</details>


---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/rowwriter.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfuFgb8fuhyPmr6qWZ:-KHfuFgb8fuhyPmr6qW_:-1473834059) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/rowwriter.go#L157)):*
> ```Go
> 	// checksum includes the key the value is associated with.
> 	ri.sentinelValue.SetBytes([]byte{})
> 	if ignoreConflicts {
> ```

consider DRYING this up
```
putFn := func(...) {
  if log.V(2) {
    ...
  }
  b.CPut(...)
}
if ignoreConflicts {
  putFn = func(...) {
    ....
  }
  b.Put(...)
}
```

---

*[sql/tablewriter.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvDT6kGXihhUjt9VA:-KHfvDT6kGXihhUjt9VB:-1102269543) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/tablewriter.go#L144)):*
> ```Go
> 	eval(insertRow parser.DTuple, existingRow parser.DTuple) (parser.DTuple, error)
> 
> 	isIdentityEvaler() bool
> ```

not new but is there only one implementation of this interface? why bother with an interface?

---

*[sql/tablewriter.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvJmiOstMv0w_anvL:-KHfvJmiOstMv0w_anvM:1389319111) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/tablewriter.go#L159)):*
> ```Go
> 	txn                   *client.Txn
> 	tableDesc             *sqlbase.TableDescriptor
> 	fastPathBatch         *client.Batch
> ```

how come a new batch is needed in this case?

---

*[sql/testdata/upsert, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvijg-P8U-yWZwDtV:-KHfvijg-P8U-yWZwDtW:-458342757) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/testdata/upsert#L48)):*
> ```
> INSERT INTO kv VALUES (13, 13), (7, 8) ON CONFLICT (k) DO NOTHING
> 
> # For better or worse, the fast path allows updating the same row twice in one
> ```

what does PG do?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-:-KHfuvxk_q_JnP4-xVyN:1210023163)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/testdata/upsert, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvijg-P8U-yWZwDtV:-KHj7IwRXaYrcRt6gCCm:-1749234366) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/testdata/upsert#L48)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what does PG do?</details>
It's unexpected to have a change in semantics that depends on an internal optimization.. Can't we put all the keys or PK values in a map and detect this case?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/testdata/upsert, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvijg-P8U-yWZwDtV:-KHrbMOIAnuZItLg-wTG:923001409) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/testdata/upsert#L48)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> It's unexpected to have a change in semantics that depends on an internal optimization.. Can't we put all the keys or PK values in a map and detect this case?</details>
Yeah, we just went through this in #6564. I think we need to continue to raise an error on conflicts, even if those conflicts are in the same statement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673)*

**Comment by paperstreet** [comment]
Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[sql/rowwriter.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfuFgb8fuhyPmr6qWZ:-KHuGO7w9qJqW4-n3wMp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/rowwriter.go#L157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> consider DRYING this up
> ```
> putFn := func(...) {
>   if log.V(2) {
>     ...
>   }
>   b.CPut(...)
> }
> if ignoreConflicts {
>   putFn = func(...) {
>     ....
>   }
>   b.Put(...)
> }
> ```</details>
Done.

---

*[sql/tablewriter.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvDT6kGXihhUjt9VA:-KHu95euaDN-6wSbxz4d:-738370381) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/tablewriter.go#L144)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not new but is there only one implementation of this interface? why bother with an interface?</details>
tl;dr Yes, this interface provides no value and should go away.

I was trying to separate the logic of how we represent and mutate structured tables from the rest of the sql stuff. Basically, if we were to change how we represent tables (like the upcoming column groups work) it lives in here, otherwise it doesn't. Unfortunately, it was a misguided effort. I realized since this went in that I should be focusing more on which code will be pushed down below the sql gateway in a distsql world and that will certainly involve evaling sql exprs somehow.

---

*[sql/tablewriter.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvJmiOstMv0w_anvL:-KHuAQtAN6ZzirmcGBcd:441615178) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/tablewriter.go#L159)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come a new batch is needed in this case?</details>
In the non-fast path case, there are a 2-3 `client.Batch`s created and run per call to flush, which is unfortunate but necessary complexity. In the fast path, the logic is much simpler, so this can work more similarly to insert/update/delete, which create a `client.Batch` in init and run it in finalize.

---

*[sql/testdata/upsert, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvijg-P8U-yWZwDtV:-KHuJDP73yoDNYQ1p2co:-868904382) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/testdata/upsert#L48)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yeah, we just went through this in #6564. I think we need to continue to raise an error on conflicts, even if those conflicts are in the same statement.</details>
Okay, I was 50/50 on this, but it sounds like there's consensus for consistency.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673)*

**Comment by RaduBerinde** [comment]
*[sql/testdata/upsert, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvijg-P8U-yWZwDtV:-KHuPM3ghi874NWkILy-:1736836957) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/testdata/upsert#L48)):*
<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> Okay, I was 50/50 on this, but it sounds like there's consensus for consistency.</details>
Nice, sounds like you don't need to do anything other than either wait for that PR to go in, or make sure the tests gets updated as part of it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673)*

**Comment by tamird** [comment]
Reviewed 2 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[sql/rowwriter.go, line 135 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHuWeJZT0wutAVeIqfx:-KHuWeJZT0wutAVeIqfy:1651046560) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2a0c9dc85a3948ab8766c73f67cd764c740d249/sql/rowwriter.go#L135)):*
> ```Go
> 	// putFn `CPut`s a key and value if conflicts should be respected, and `Put`s
> 	// them otherwise. logValue is used to pretty print the value.
> 	putFn := func(key *roachpb.Key, value interface{}, logValue interface{}) {
> ```

note that this currently allocates - you could change both these `Fn`s to actual functions (and pass in the batch) to avoid the allocation.

---

*[sql/tablewriter.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvDT6kGXihhUjt9VA:-KHuWnRa93KwvyGNV4bJ:-2133011968) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/tablewriter.go#L144)):*
<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> tl;dr Yes, this interface provides no value and should go away.
> 
> I was trying to separate the logic of how we represent and mutate structured tables from the rest of the sql stuff. Basically, if we were to change how we represent tables (like the upcoming column groups work) it lives in here, otherwise it doesn't. Unfortunately, it was a misguided effort. I realized since this went in that I should be focusing more on which code will be pushed down below the sql gateway in a distsql world and that will certainly involve evaling sql exprs somehow.</details>
OK. Does that deserve a TODO or an issue?

---

*[sql/tablewriter.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvJmiOstMv0w_anvL:-KHuWBktJ6ZI7AHolLYo:-750544633) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/tablewriter.go#L159)):*
<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> In the non-fast path case, there are a 2-3 `client.Batch`s created and run per call to flush, which is unfortunate but necessary complexity. In the fast path, the logic is much simpler, so this can work more similarly to insert/update/delete, which create a `client.Batch` in init and run it in finalize.</details>
Bummer. Could you add a comment about this oddity?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673)*

**Comment by paperstreet** [comment]
Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[sql/rowwriter.go, line 135 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHuWeJZT0wutAVeIqfx:-KHueFV1a2YHoFCf_vYW:-1366381672) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2a0c9dc85a3948ab8766c73f67cd764c740d249/sql/rowwriter.go#L135)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> note that this currently allocates - you could change both these `Fn`s to actual functions (and pass in the batch) to avoid the allocation.</details>
Oh cool. Done.

I should have rerun the benchmarks and caught that myself

---

*[sql/tablewriter.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvDT6kGXihhUjt9VA:-KHudPb1uGoOl22sjVjb:2133586090) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/tablewriter.go#L144)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK. Does that deserve a TODO or an issue?</details>
TODOed

---

*[sql/tablewriter.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvJmiOstMv0w_anvL:-KHudP0W-pBoxVmRzDfz:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/tablewriter.go#L159)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Bummer. Could you add a comment about this oddity?</details>
Done.

---

*[sql/testdata/upsert, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHfvijg-P8U-yWZwDtV:-KHu_BA2yToxq5t8_k-k:1028595652) ([raw file](https://github.com/cockroachdb/cockroach/blob/7deab6a92a8c888b4bab6fbe032e048bd17219ae/sql/testdata/upsert#L48)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Nice, sounds like you don't need to do anything other than either wait for that PR to go in, or make sure the tests gets updated as part of it?</details>
That PR would only catch this when the blind put optimization is used, which (IIUC) it's not guaranteed here, so I need the explicit check added in r2.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673)*

**Comment by tamird** [comment]
:lgtm:

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Ah, makes sense.</details>


---

Reviewed 1 of 3 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/rowwriter.go, line 135 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-KHuWeJZT0wutAVeIqfx:-KHuuUI-fmAa4tjPyygi:-461299781) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2a0c9dc85a3948ab8766c73f67cd764c740d249/sql/rowwriter.go#L135)):*
<details><summary><i>Previously, paperstreet (Dan Harrison) wrote…</i></summary>
> Oh cool. Done.
> 
> I should have rerun the benchmarks and caught that myself</details>
:+1:

FYI these methods can be used to DRY up quite a bit of code in this package now, might make a good new contributor task.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6673#-:-KHuujmJ--Fn9EgIezI0:1417088721)*
